lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -2565,14 +2554,52 @@ void TFDeabstraction::doIt() {
   // Now that we've promoted all the allocations in the way of our dataflow,
   // go through and propagate any tuple/struct values that are in the way of
   // our analysis.
-  propagateSSAValues();
+  propagateSSAValues(tensorOps);
 
   logCurrentState("After propagateSSAValues", /*detailed*/ true);
 
   // Canonicalize attribute arguments, check that they have constants,
   // flatten array attributes, and form graph_op instructions.
   checkAttributesAndFormGraphOps();
 
+  logCurrentState("After checkAttributesAndFormGraphOps", /*detailed*/ true);
+
+  // After graph_op instructions are formed, we have some opportunities for
+  // optimizing code that would result in better code for partitioning. We do
+  // them now.
+  //
+  // e.g., the inlining of functions (like allocateUninitializedArray) may
+  // result in the following code snippet in a function:
+  //     %9 = alloc_stack $Int32
```

Comment:
can you provide an example on why we should care about optimizing such scalar (non-tensor) insts here? is it for tensorflow convention function?

Reply:
the larger question is that if we start applying more optimizations that used to belong only to the optimization pipeline, could that be "slippery slope" in that we might keep adding more such logic to the DA pass over time?

It'd be nice if we could try and make the DA pass compose with (and complimentary to) the other compiler passes as much as possible.

Reply:
This specific optimization mainly avoids the introduction of sends and receives by eliminating redundant allocations and creating an SSA value for attribute values.  This is actually quite similar to promoteToSSA function. (May be it is sufficient to call promoteToSSA again. Let me try it.)

When working on this PR, I simply tried to avoid introducing new sends and receives as much as possible in our tests. As a result of that approach, I identified that these optimizations were necessary. Given that we won't be running the optimization pipeline for tensor-flow convention functions in -Onone mode and still be able to extract graphs, we will have to perform some of the transformations here.

Reply:
I have the same question as Mingsheng.  with partitioning running after the performance optimizer, it seems like these sorts of things should already be done.  Why is this necessary?  What is the bad thing that happens if we don't call into "optimizeMemoryAllocations" here?  Is this a phase ordering issue that can be resolved some other way?